plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/ProjectContextProvider.kt

```diff
@@ -0,0 +1,355 @@
+        logger.info { "project context: getting usage for ${project.name} on port ${encoderServer.port}" }
+        val url = URL("http://localhost:${encoderServer.port}/getUsage")
+        val connection = url.openConnection() as HttpURLConnection
+        setConnectionProperties(connection)
+        val responseCode = connection.responseCode
+
+        logger.info { "project context getUsage response code: $responseCode for ${project.name} " }
+        val responseBody = if (responseCode == 200) {
+            connection.inputStream.bufferedReader().use { reader -> reader.readText() }
+        } else {
+            ""
+        }
+        connection.disconnect()
+        try {
+            val parsedResponse = mapper.readValue<Usage>(responseBody)
+            return parsedResponse
+        } catch (e: Exception) {
+            logger.warn { "error parsing query response ${e.message}" }
+            return null
+        }
+    }
+
+    fun updateIndex(filePath: String) {
+        logger.info { "project context: updating index for $filePath on port ${encoderServer.port}" }
+        val url = URL("http://localhost:${encoderServer.port}/updateIndex")
+        val payload = UpdateIndexRequestPayload(filePath)
+        val payloadJson = mapper.writeValueAsString(payload)
+        val encrypted = encoderServer.encrypt(payloadJson)
+        with(url.openConnection() as HttpURLConnection) {
+            setConnectionProperties(this)
+            setConnectionRequest(this, encrypted)
+            val responseCode = responseCode
+            logger.debug { "project context update index response code: $responseCode for $filePath" }
+            return
+        }
+    }
+
+    private fun setConnectionTimeout(connection: HttpURLConnection) {
+        connection.connectTimeout = 5000 // 5 seconds
+        connection.readTimeout = 10000 // 10 second
+    }
+
+    private fun setConnectionProperties(connection: HttpURLConnection) {
+        connection.requestMethod = "POST"
+        connection.setRequestProperty("Content-Type", "text/plain")
+        connection.setRequestProperty("Accept", "text/plain")
+    }
+
+    private fun setConnectionRequest(connection: HttpURLConnection, payload: String) {
+        connection.doOutput = true
+        connection.outputStream.use { outputStream ->
+            OutputStreamWriter(outputStream).use { writer ->
+                writer.write(payload)
+            }
+        }
+    }
+
+    private fun willExceedPayloadLimit(currentTotalFileSize: Long, currentFileSize: Long): Boolean {
+        val maxSize = CodeWhispererSettings.getInstance().getProjectContextIndexMaxSize()
+        return currentTotalFileSize.let { totalSize -> totalSize > (maxSize * 1024 * 1024 - currentFileSize) }
+    }
+
+    private fun isBuildOrBin(filePath: String): Boolean {
+        val regex = Regex("""[/\\](bin|build|node_modules|env|\.idea)[/\\]""", RegexOption.IGNORE_CASE)
+        return regex.find(filePath) != null
+    }
```

Comment:
we can ask the IDE what is supposed to be ignored

Reply:
ack this one can be a followup item